lumeer-storage/lumeer-storage-mongodb/src/main/java/io/lumeer/storage/mongodb/dao/system/MongoUserDao.java

```diff
@@ -0,0 +1,195 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.lumeer.storage.mongodb.dao.system;
+
+import static io.lumeer.storage.mongodb.util.MongoFilters.idFilter;
+
+import io.lumeer.api.model.User;
+import io.lumeer.storage.api.dao.UserDao;
+import io.lumeer.storage.api.exception.StorageException;
+import io.lumeer.storage.mongodb.MongoUtils;
+import io.lumeer.storage.mongodb.codecs.UserCodec;
+import io.lumeer.storage.mongodb.model.MongoUser;
+
+import com.mongodb.MongoException;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.FindOneAndUpdateOptions;
+import com.mongodb.client.model.IndexOptions;
+import com.mongodb.client.model.Indexes;
+import com.mongodb.client.model.ReturnDocument;
+import com.mongodb.client.model.UpdateOptions;
+import com.mongodb.client.model.Updates;
+import com.mongodb.client.result.UpdateResult;
+
+import org.bson.BsonDocument;
+import org.bson.Document;
+import org.bson.conversions.Bson;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.RequestScoped;
+
+@RequestScoped
+public class MongoUserDao extends SystemScopedDao implements UserDao {
+
+   private static final String COLLECTION_NAME = "users";
+   private static final String ELEMENT_NAME = "group";
+
+   @PostConstruct
+   public void checkRepository() {
+      if (database.getCollection(COLLECTION_NAME) == null) {
+         createUsersRepository();
+      }
+   }
+
+   public void createUsersRepository() {
+      database.createCollection(databaseCollectionName());
+
+      MongoCollection<Document> userCollection = database.getCollection(databaseCollectionName());
+      userCollection.createIndex(Indexes.ascending(UserCodec.EMAIL), new IndexOptions().unique(true));
+   }
+
+   public void deleteUsersRepository() {
+      database.getCollection(databaseCollectionName()).drop();
+   }
+
+   @Override
+   public User createUser(final String organizationId, final User user) {
+      MongoUser existingUser = databaseCollection().find(Filters.eq(UserCodec.EMAIL, user.getEmail())).first();
+      if (existingUser != null) {
+         return upsertUserGroupsForOrganization(organizationId, existingUser.getId(), user);
+      }
+      return createNewUser(organizationId, user);
```

Comment:
Why do we have all this "business logic" here? DAO layer should be really stupid and responsible only for storing and retrieving data. If we don't follow these rules which we set, we will eventually end up with the same spaghetti code we once had here.

You can clearly see that something is wrong just by looking at the `UserFacade` which doesn't really do anything useful (except for checking the permissions) and just calls the DAO methods.

Reply:
It's more complicated than you think. Creating users is not always creating, because when user already exist in db, we just add information that he is in some organization (by its id). So first time it is real create in db, and next time we do update. So who should be responsible for knowing structure of how users are stored in db?

Reply:
Creating user on the DAO layer should be only about creating the database record. Creating user on the facade and service layer should be a more complex operation (what front-end expects and what you have implemented in the DAO).

The only thing that needs to be done is moving the content this method to the facade and replace database calls by DAO method calls. If the user exists, you will update it. If not, you will creat it. Is that clear?

I'm just trying to avoid the situation we had here half a year ago. If we don't follow these simple rules, after a few such changes we will be back there and every change will be a pain again.

Reply:
What I'm trying to tell you is that there does not always needs to be 1-to-1 mapping between the DAO and facade methods. They can sometimes do something slightly different. Otherwise, we would not need these distinct layers.